odoo/addons/base/tests/test_qweb.py

```diff
@@ -1762,6 +1702,977 @@ def test_space_remove_technical_all(self):
+                    <article>1</article>
+                    <div>cache: 101</div>
+                </section>
+            </root>
+        """
+        self.assertEqual(etree.fromstring(render), etree.fromstring(result), 'rendering 1 (1 != 101: cached t-set should is applied on first rendering)')
+
+        render = IrQweb._render(template_page.id, {
+            'cache_id': 1,
+            'counter': 2,
+        })
+        result = """
+            <root>
+                <section>
+                    <article>2</article>
+                    <div>cache: 101</div>
+                </section>
+            </root>
+        """
+        self.assertEqual(etree.fromstring(render), etree.fromstring(result), 'rendering 2 (2 != 102: cached t-set should be applied the template part are rendered every time)')
+
+        render = IrQweb._render(template_page.id, {
+            'cache_id': 3,
+            'counter': 3,
+        })
+        result = """
+            <root>
+                <section>
+                    <article>3</article>
+                    <div>cache: 103</div>
+                </section>
+            </root>
+        """
+        self.assertEqual(etree.fromstring(render), etree.fromstring(result), 'rendering 3 (3 != 103: cached t-set should applied because the new cache key is created)')
+
+    def test_13_render_xml_cache_with_t_set_in_cache(self):
+        template_page = self.env['ir.ui.view'].create({
+            'name': "template_page",
+            'type': 'qweb',
+            'arch': """
+                <t t-name="template_page">
+                    <root>
+                        <section t-cache="(cache_id,)">
+                            <t t-set="counter" t-value="counter + 100"/>
+                            <article t-nocache=""><t t-out="counter"/></article>
+                            <div>cache: <t t-out="counter"/></div>
+                        </section>
+                        <div>out of cache: <t t-out="counter"/></div>
+                    </root>
+                </t>
+            """
+        })
+        IrQweb = self.env['ir.qweb'].with_context(use_qweb_t_cache=True)
+
+        render = IrQweb._render(template_page.id, {
+            'cache_id': 1,
+            'counter': 1,
+        })
+        result = """
+            <root>
+                <section>
+                    <article>1</article>
+                    <div>cache: 101</div>
+                </section>
+                <div>out of cache: 101</div>
+            </root>
+        """
+        self.assertEqual(etree.fromstring(render), etree.fromstring(result), 'rendering 1')
+
+        render = IrQweb._render(template_page.id, {
+            'cache_id': 1,
+            'counter': 2,
+        })
+        result = """
+            <root>
+                <section>
+                    <article>2</article>
+                    <div>cache: 101</div>
+                </section>
+                <div>out of cache: 2</div>
```

Comment:
That's weird: the fact that the template was in cache impacts the elements not int cache. I would expect to either see 102 or 1 at Line 2533 (because the original value has been passed instead of the one computed inside the t-set of the cached section)

Reply:
The `t-nocache` uses the values of the root (without modification), the `t-cache` does not copy the `values` and therefore modifies the dictionary. I'll test on runbot that with each `t-cache`, the values dictionary is copied.